File: java/src/main/java/vsdk/toolkit/io/bin/reader/BinaryModelPatchRecordData.java
package vsdk.toolkit.io.bin.reader;

import vsdk.toolkit.common.ColorRgb;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.skin.Patch;

public class BinaryModelPatchRecordData {
    public int id;
    public int twinIndex;
    public int numberOfVertices;
    public int[] vertexIndices;
    public boolean hasBoundingBox;
    public float[] boundingBoxCoordinates;
    public Vector3D normal;
    public float planeConstant;
    public float tolerance;
    public float area;
    public Vector3D midPoint;
    public boolean hasJacobian;
    public float jacobianA;
    public float jacobianB;
    public float jacobianC;
    public float directPotential;
    public int dominantIndex;
    public boolean omit;
    public byte flags;
    public ColorRgb color;
    public int materialIndex;
    public boolean hasRadianceData;

    public BinaryModelPatchRecordData() {
        id = 0;
        twinIndex = -1;
        numberOfVertices = 0;
        vertexIndices = new int[Patch.MAXIMUM_VERTICES_PER_PATCH];
        for (int i = 0; i < Patch.MAXIMUM_VERTICES_PER_PATCH; i++) {
            vertexIndices[i] = -1;
        }
        hasBoundingBox = false;
        boundingBoxCoordinates = new float[6];
        for (int i = 0; i < 6; i++) {
            boundingBoxCoordinates[i] = 0.0f;
        }
        normal = new Vector3D();
        planeConstant = 0.0f;
        tolerance = 0.0f;
        area = 0.0f;
        midPoint = new Vector3D();
        hasJacobian = false;
        jacobianA = 0.0f;
        jacobianB = 0.0f;
        jacobianC = 0.0f;
        directPotential = 0.0f;
        dominantIndex = 0;
        omit = false;
        flags = 0;
        color = new ColorRgb();
        materialIndex = -1;
        hasRadianceData = false;
    }
}

File: typescript/src/main/vsdk/toolkit/io/bin/reader/BinaryModelPatchRecordData.ts
import { ColorRgb } from "../../../common/ColorRgb";
import { Vector3D } from "../../../common/linealAlgebra/Vector3D";
import { Patch } from "../../../skin/Patch";

export class BinaryModelPatchRecordData {
  public id: number;
  public twinIndex: number;
  public numberOfVertices: number;
  public vertexIndices: number[];
  public hasBoundingBox: boolean;
  public boundingBoxCoordinates: number[];
  public normal: Vector3D;
  public planeConstant: number;
  public tolerance: number;
  public area: number;
  public midPoint: Vector3D;
  public hasJacobian: boolean;
  public jacobianA: number;
  public jacobianB: number;
  public jacobianC: number;
  public directPotential: number;
  public dominantIndex: number;
  public omit: boolean;
  public flags: number;
  public color: ColorRgb;
  public materialIndex: number;
  public hasRadianceData: boolean;

  public constructor() {
    this.id = 0;
    this.twinIndex = -1;
    this.numberOfVertices = 0;
    this.vertexIndices = new Array<number>(Patch.MAXIMUM_VERTICES_PER_PATCH).fill(-1);
    this.hasBoundingBox = false;
    this.boundingBoxCoordinates = new Array<number>(6).fill(0.0);
    this.normal = new Vector3D();
    this.planeConstant = 0.0;
    this.tolerance = 0.0;
    this.area = 0.0;
    this.midPoint = new Vector3D();
    this.hasJacobian = false;
    this.jacobianA = 0.0;
    this.jacobianB = 0.0;
    this.jacobianC = 0.0;
    this.directPotential = 0.0;
    this.dominantIndex = 0;
    this.omit = false;
    this.flags = 0;
    this.color = new ColorRgb();
    this.materialIndex = -1;
    this.hasRadianceData = false;
  }
}
